docs/50_writing_tests.md

```diff
@@ -0,0 +1,463 @@
+
+In the test project
+* create a new folder for the device under tst/Devices. This is for anything you define for testing this device.
+* add a new class with header file and optionally .cpp file to the project in that folder. Name the cass DeviceNameTestData.
+* edit the header file to inherit the class from the testdata abstract class appropriate to the device type, i.e. BikeTestData, RowerTestData, EllipticalTestData, TreadmillTestData.
+* have this subclass' constructor pass a unique test name to the superclass.
+
+The tests are not organised around real devices that are handled, but the bluetoothdevice class that handles them - the "driver" of sorts.
+
+You need to provide the following:
+- patterns for valid names (e.g. equals a value, starts with a value, case sensitivity, specific length)
+- invalid names to ensure the device is not identified when the name is invalid
+- configuration settings that are required for the device to be detected
+- invalid configurations to test that the device is not detected, e.g. when it's disabled in the settings, but the name is correct
+- exclusion devices: if a device with the same name but of a higher priority type is detected, this devivce should not be detected
+- valid and invalid QBluetoothDeviceInfo configurations, e.g. to check the device is only detected when the manufacturer data is set correctly, or certain services are availble or not.
+
+## Tools in the Test Framework
+
+### TestSettings
+
+The detection of many devices depends on settings that are accessed programmatically using the QSettings class and the constants in the QZSettings namespace. The TestSettings class stores a QSettings object with what is intended to be a unique application and organisation name, to keep the configuration it represents seperate from others in the system. It also makes the stored QSettings object the default by setting the QCoreApplication's organisation and application names to those of the QSettings object. The original values are restored by calling the deactivate() function or on object destruction.
+
+i.e. a test will 
+* apply a configuration from a TestSettings object
+* perform device detection
+* use the TestSettings object to restore the previous settings either directly or by letting its destructor be called.
+
+## Writing a device detection test
+
+Because of the way the TestData classes currently work, it may be necessary to define multiple test data classes to cover the various cases.
+For example, if any of a list of names is enough to identify a device, or another group of names but with a certain service in the bluetooth device info, that will require multiple classes.
+
+### Recognition by Name
+
+Consider the detection code for the Domyos Bike:
+
+```
+ } else if (b.name().startsWith(QStringLiteral("Domyos-Bike")) &&
+                       !b.name().startsWith(QStringLiteral("DomyosBridge")) && !domyosBike && filter) {
+
```

Comment:
@cagnulein Perhaps this expression in bluetooth.cs requires an update. If the name starts with Domyos-Bike it will never start with DomyosBridge.

Reply:
yes that's right. Domyos-bike must use the domyosbike class, isn't it @drmason789  ?